Enhanced debugger for godot 3.0

Geequlim opened the pull request:
Some improvements for the debugger of godot 3.0. As disscused with @reduz in #10891.

The PR will take the enhanced debugger like the cuurent debugger 2.1.4 to the master branch.

- [x] Make it possiable to access members and constants from all kind of script languages.
- [x] Send script varialbes and constants when inspector remote objects.
- [x] Improved property editor to interact with remote objects.
- [x] Merge debugger inspector with normal inspector.
- [x] Merge live scene tree panel with normal scene tree.
- [x] Implement a ~~better~~ dictionary editor ~~than the 2.1 version~~.
   * It is hard to implement that with current PropertyEditor as discussed in #12015. So I will take the 2.1.4  version with some improvements.

reduz commented:
Did you see my proposal in the 2.1 issue? Seems like it from your two last checkboxes, but making sure :P
I would also add that live editing should probably be enabled by default too.

Geequlim commented:
![2017-10-15 00-38-34](https://user-images.githubusercontent.com/6964556/31577469-be53ec02-b141-11e7-90b4-842ce5bb8c0b.png)
@reduz I have implemented the inspector to show and edit remote objects now but the live scene tree still stays here.

reduz commented:
looking great!

Geequlim commented:
![2017-10-15 00-51-40](https://user-images.githubusercontent.com/6964556/31577540-18daa606-b143-11e7-8ebb-6a1ad0d938c8.png)
Why I didn't move the `Live Scene Tree` to the normal Scene Tree dock is I didn't find a good way to manage the tree with the debugger but it is contained in another place.

That would make the code more mess :confused:

reduz commented:
I can help you move it when I get back home on Tuesday

On Oct 14, 2017 1:56 PM, "Geequlim" <notifications@github.com> wrote:

> [image: 2017-10-15 00-51-40]
> <https://user-images.githubusercontent.com/6964556/31577540-18daa606-b143-11e7-8ebb-6a1ad0d938c8.png>
> Why I didn't move the Live Scene Tree to the normal Scene Tree dock is I
> didn't find a good way to manage the tree with the debugger but it is
> contained in another place.
>
> That would make the code more mess 😕
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/11940#issuecomment-336648414>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF-Z20LJHGXxDMi4XMXw3EwlJ3YSVh3Tks5ssOfCgaJpZM4PxryO>
> .
>

Geequlim committed:
Abstract some method for script system

Geequlim committed:
Send script members, contants and globals to debugger
Remove remote inspector panel

Geequlim committed:
Better supported for remote object editing with the inspector with a simple dictionary editor implement

Geequlim commented:
The commits of this PR is not event to my [origin branch: debugger](https://github.com/GodotExplorer/godot/commits/debugger) after I `git push --force` to it.
Could you please give me a help about this?

Geequlim commented:
Seems the github delayed to synchronize the commit from the origin branch.

Geequlim commented:
@reduz Are you going to have a review and mege this before alpha2 ?

reduz commented:
Hi @Geequlim I was planning this week to add the tools so you can properly implement: 
"Merge live scene tree panel with normal scene tree.", in hopes of having it implemented before beta. It seems you completed the work for everything else, so will try to have this working ASAP

Geequlim commented:
@reduz Great!
I hope we can finish  this work within this week so it can be merge before the 3.0 BETA . Then I have enough time to fix bugs before the 3.0 stable.

MarianoGnu commented:
why not adding the remote scene tree to a dock in the same slot as Scene and make it active when debugging? (or add a setting to allow enable it automatically)
Other way (here comes a big complain, because of extreme and unneded complexity) is adding a "remote scene" to scene tabs, with the X button not enabled, then you can see the scene in realtime while the game is runing (you have a duplicate of the nodes with the same property values, but scripts disabled)

reduz commented:
@Geequlim alright, added the support you were missing. You can acess SceneTreeDock by using 
```
EditorNode::get_singleton()->get_scene_tree_dock()
```

so, basically add your remote tree editor with

```
EditorNode::get_singleton()->get_scene_tree_dock()->add_remote_tree_editor(editor)
```

When debug begins, show it by doing:

```
EditorNode::get_singleton()->get_scene_tree_dock()->show_remote_tree()
```

and when debug ends, hide it by doing:

```
EditorNode::get_singleton()->get_scene_tree_dock()->hide_remote_tree()
```

and the remote debug menu can go away!

Geequlim committed:
Move the remote scene tree to the scene tree dock.
Ignore all script constants in the global section of the breakpoint stack.
Check property size before send to avoid too large of data be sent.
Fix crash while clear the remote objects from the debugger.

Geequlim commented:
All features are implemented now.
It is ready for merging.

akien-mga commented:
Some screenshots to build up the hype? :D

Geequlim commented:
@akien-mga  Here we go!

* Explore remote member variables with dictionary support in breakpoint stacks.
![remote member variables](https://user-images.githubusercontent.com/6964556/32884714-3174f51e-caf6-11e7-9141-49e4f8332f85.png)

* Live explore the remote scene tree
![Live explore the remote scene tree](https://user-images.githubusercontent.com/6964556/32885032-3d069eb8-caf7-11e7-9fc8-8732bf91e299.gif)

* Edit remote objects from the inspector directly
![ Edit remote objects](https://user-images.githubusercontent.com/6964556/32885129-8152418a-caf7-11e7-93f2-e7bd32b51bdc.gif)

n-pigeon commented:
This is excelsior!

reduz commented on core/script_debugger_remote.cpp:
```diff
@@ -532,56 +553,77 @@ void ScriptDebuggerRemote::_send_object_id(ObjectID p_id) {
 	if (!obj)
 		return;
 
-	List<PropertyInfo> pinfo;
```
I actually like my code for handling requesting objects better. It's newer than yours, fixes a few bugs,  sends Object IDs and can handle when an object is too big to send.

Geequlim replied:
I remember you have changed the limit to 8MB it should be enough for most objects. I will check your code again and make it more reliable with this. But it is necessary to change something to send more usable informations like script constant values and member variables of script instance.

reduz replied:
member variables of script instance should already be obtainable with get_property_list to the object, they have the usage flag: PROPERTY_USAGE_SCRIPT_VARIABLE , so no extra code is needed.

reduz commented on core/script_debugger_remote.cpp:
```diff
@@ -231,11 +252,11 @@ void ScriptDebuggerRemote::debug(ScriptLanguage *p_script, bool p_can_continue)
 
-				{ //locals
-					packet_peer_stream->put_var(locals.size());
+				{ //globals
```
The idea of sending globals is fine, but in GDScript, globals are constants and there are a ton of them. Sending them every time seems like too much.

Geequlim replied:
Infact only autoloads will be sent in GDScript.

reduz commented on editor/script_editor_debugger.cpp:
```diff
@@ -399,55 +425,57 @@ void ScriptEditorDebugger::_parse_message(const String &p_msg, const Array &p_da
-			inspected_object->prop_list.clear();
-			inspected_object->prop_values.clear();
+		bool is_new_object = false;
+		if (remote_objects.has(id)) {
```
Ah, I like this, makes the history work.

reduz commented on editor/script_editor_debugger.cpp:
```diff
@@ -399,55 +425,57 @@ void ScriptEditorDebugger::_parse_message(const String &p_msg, const Array &p_da
 		}
 
-		for (int i = 0; i < prop_count; i++) {
+		for (int i = 0; i < properties.size(); i++) {
```
Please check my code for sending objects, it fixes a few bugs since 2.1 regarding objects being too big that may crash the debugger, so you may want to use it, or part of it.

akien-mga commented:
Sorry @Geequlim, I did a slight refactoring of the gdscript module in #12969 so you'll need to rebase this PR.

Geequlim commented:
@akien-mga Rebased to latest master.

@reduz I try to use your code to send properties but I got some problems
* Constant values can't be send.
* Some of script members are not listed in the result even with flag `PROPERTY_USAGE_SCRIPT_VARIABLE`
* The debugger stopped with parsing the received  data with [script_editor_debugger.cpp:1042](https://github.com/godotengine/godot/blob/master/editor/script_editor_debugger.cpp#L1042) and [script_editor_debugger.cpp:1055](https://github.com/godotengine/godot/blob/master/editor/script_editor_debugger.cpp#L1055)

So I keeped the code of mine for send objects with your size checking logic for properties.

reduz commented:
@Geequlim If you want to keep your code but use the size checking it's fine

reduz commented:
@Geequlim I don't mind, but please make sure to use my code to check that no buffer overrides exist, otherwise the editor will crash.

reduz commented:
@Geequlim also, cant seem to drag values in the inspector

Geequlim commented:
@reduz I have take the size checking code inside the PR before merged. But I’m not sure it is safe enough as the size checking for each property to the max size of the stream peer as it did before.
